src/crystal/compiler_rt/shift.cr

```diff
@@ -0,0 +1,53 @@
+# :nodoc:
+# Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/lib/builtins/ashlti3.c
+#
+# Returns: a << b
+# Precondition:  0 <= b < bits_in_tword
+fun __ashlti3(a : Int128, b : Int32) : Int128
+  low, high = a.unsafe_as(Tuple(UInt64, Int64))
+  if b >= 64
+    low, high = 0u64, (low << (b &- 64)).to_i64!
+  elsif b == 0
+    return a
+  else
+    low, high = low << b, (high << b) | (low >> (64 &- b))
+  end
+
+  {low, high}.unsafe_as(Int128)
```

Comment:
Same here, using `Tuple(_, _).unsafe_as(_)` feels like abusing the method

How it's done in other parts of `compiler-rt`:

https://github.com/crystal-lang/crystal/blob/100391991856c53d31e3f712ff8ab5197d687146/src/crystal/compiler_rt/divmod128.cr#L178

https://github.com/crystal-lang/crystal/blob/100391991856c53d31e3f712ff8ab5197d687146/src/crystal/compiler_rt/multi3.cr#L18

Reply:
Then I guess this can also be considered resolved